packages/ws/src/ws/WebSocketShard.ts

```diff
@@ -0,0 +1,515 @@
+				);
+			}
+		}
+
+		const url = `${data.url}?${params.toString()}`;
+		this.debug([`Connecting to ${url}`]);
+		const connection = new WebSocket(url, { handshakeTimeout: this.strategy.options.handshakeTimeout ?? undefined })
+			/* eslint-disable @typescript-eslint/no-misused-promises */
+			.on('message', this.onMessage.bind(this))
+			.on('error', this.onError.bind(this))
+			.on('close', this.onClose.bind(this));
+		/* eslint-enable @typescript-eslint/no-misused-promises */
+
+		connection.binaryType = 'arraybuffer';
+		this.connection = connection;
+
+		this.status = WebSocketShardStatus.Connecting;
+
+		await this.waitForEvent(WebSocketShardEvents.Hello, this.strategy.options.helloTimeout);
+
+		const session = this.session ?? (await this.strategy.retrieveSessionInfo(this.id));
+		if (session?.shardCount === this.strategy.options.shardCount) {
+			await this.resume(session);
+		} else {
+			await this.identify();
+		}
+	}
+
+	public async destroy(options: WebSocketShardDestroyOptions = {}) {
+		if (this.status === WebSocketShardStatus.Idle) {
+			throw new Error('Tried to destroy a shard that was idle');
+		}
+
+		if (!options.code) {
+			options.code = options.recover === WebSocketShardDestroyRecovery.Resume ? 4200 : 1000;
+		}
+
+		this.debug([
+			'Destroying shard',
+			`Reason: ${options.reason ?? 'none'}`,
+			`Code: ${options.code}`,
+			`Recover: ${options.recover === undefined ? 'none' : WebSocketShardDestroyRecovery[options.recover]!}`,
+		]);
+
+		// Reset state
+		this.isAck = true;
+		if (this.heartbeatInterval) {
+			clearInterval(this.heartbeatInterval);
+		}
+		this.lastHeartbeatAt = -1;
+
+		// Clear session state if applicable
+		if (options.recover !== WebSocketShardDestroyRecovery.Resume && this.session) {
+			this.session = null;
+			await this.strategy.updateSessionInfo(this.id, null);
+		}
+
+		if (this.connection && (this.connection.readyState === OPEN || this.connection.readyState === CONNECTING)) {
+			this.connection.close(options.code, options.reason);
+		}
+
+		this.status = WebSocketShardStatus.Idle;
+
+		if (options.recover !== undefined) {
+			return this.connect();
+		}
+	}
+
+	private async waitForEvent(event: WebSocketShardEvents, timeoutDuration?: number | null) {
+		this.debug([`Waiting for event ${event} for ${timeoutDuration ? `${timeoutDuration}ms` : 'indefinitely'}`]);
+		const controller = new AbortController();
+		const timeout = timeoutDuration ? setTimeout(() => controller.abort(), timeoutDuration).unref() : null;
+		await once(this, event, { signal: controller.signal });
+		if (timeout) {
+			clearTimeout(timeout);
+		}
+	}
+
+	public async send(payload: GatewaySendPayload) {
+		if (!this.connection) {
+			throw new Error("WebSocketShard wasn't connected");
+		}
+
+		if (this.status !== WebSocketShardStatus.Ready && !ImportantGatewayOpcodes.has(payload.op)) {
+			throw new Error('Tried to send a message before the shard was ready');
```

Comment:
This should *not* throw an error at all. If a payload is requested to be sent (think an interval to set presence because Discord's presences sometimes die) it should be enqueued till the shard is ready